### PR TITLE
batch-system: add benchmark

### PR DESCRIPTION
**Pull request by BusyJay**
###  What have you changed?

Add several benchmark cases to batch system. It will be used to measure the future improvement.

###  What is the type of the changes?

- Engineering

###  How is the PR tested?

- Unit test
- Integration test
- Manual test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No.

###  Does this PR affect `tidb-ansible`?

No.

###  Benchmark result if necessary (optional)

Following result comes from benchmark on my desktop:
```
fairness                time:   [36.352 us 37.078 us 37.817 us]
imbalance               time:   [2.3132 ms 2.3597 ms 2.3929 ms]
spawn_many              time:   [4.9201 ms 4.9222 ms 4.9240 ms]
router::send            time:   [89.818 ns 89.913 ns 90.005 ns]
```

### CODE REVIEW

**Comment by BusyJay** [comment]
@overvenus @5kbpers @gengliqi PTAL

File: components/batch-system/benches/router.rs [hunk 1]
```diff
@@ -1,85 +1,24 @@
...
+    system.spawn("test".to_owned(), Builder::new());
+    let (normal_tx, normal_fsm) = Runner::new(100000);
     let normal_box = BasicMailbox::new(normal_tx, normal_fsm);
     router.register(1, normal_box);
 
-    b.iter(|| {
-        router.send(1, ()).unwrap();
+    c.bench_function("router::send", |b| {
+        b.iter(|| {
+            router.send(1, Message::Loop(0)).unwrap();
```
Comment (overvenus) [inline components/batch-system/benches/router.rs]: Is it possible for OOM?

Comment (BusyJay) [reply components/batch-system/benches/router.rs]: It's possible but never happen on my 16G desktop.

**Comment by siddontang** [comment]
PTAL @nrc

File: components/batch-system/src/test_runner.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+use std::borrow::Cow;
+use std::sync::{Arc, Mutex};
+use tikv_util::mpsc;
+
+pub enum Message {
+    Loop(usize),
+    Callback(Box<dyn FnOnce(&mut Runner) + Send + 'static>),
+}
+
+pub struct Runner {
```
Comment (nrc) [inline components/batch-system/src/test_runner.rs]: Could you add some docs for what this is doing and how? And how the various fields are used please?

File: components/batch-system/src/test_runner.rs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    is_stopped: bool,
+    recv: mpsc::Receiver<Message>,
+    mailbox: Option<BasicMailbox<Runner>>,
+    pub sender: Option<mpsc::Sender<()>>,
+    res: usize,
```
Comment (nrc) [inline components/batch-system/src/test_runner.rs]: Could you use the full word here please? "res" is ambiguous (result, response, etc)

File: components/batch-system/benches/batch-system.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            }
+            for _ in 0..ID_LIMIT {
+                rx.recv().unwrap();
+            }
+        })
+    });
+    system.shutdown();
+}
+
+fn bench_imbalance(c: &mut Criterion) {
```
Comment (nrc) [inline components/batch-system/benches/batch-system.rs]: Could these top level functions get some docs to describe what is being tested/benchmarked?

File: components/batch-system/benches/batch-system.rs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    let (tx2, rx2) = std::sync::mpsc::channel();
+    c.bench_function("fairness", |b| {
+        b.iter(|| {
+            for _ in 0..10 {
+                for id in 4..6 {
+                    router.send(id, Message::Loop(10)).unwrap();
+                }
+            }
+            for id in 4..6 {
```
Comment (nrc) [inline components/batch-system/benches/batch-system.rs]: I'm a bit confused by these ids - are these two bench functions meant to be run at the same time? Do they interact? Is it deliberate or coincidence that the sets of ids are disjoint? I think that if there is some relation, then it needs to be documented

**Pull request merged**